File: customer-java-resttemplate/src/main/java/com/prez/ws/CustomerWSClient.java
package com.prez.ws;

import static org.springframework.web.util.UriComponentsBuilder.fromHttpUrl;

import com.prez.ws.model.GetCustomerWSResponse;
import java.net.URI;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.cloud.sleuth.annotation.NewSpan;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestOperations;

public class CustomerWSClient {

  private static final Logger LOGGER = LoggerFactory.getLogger(CustomerWSClient.class);

  private final RestOperations customerWebClient;
  private final CustomerWSProperties configuration;


  public CustomerWSClient(CustomerWSProperties configuration, RestOperations customerWebClient) {
    this.configuration = configuration;
    this.customerWebClient = customerWebClient;
  }

  @NewSpan("getCustomers")
  public Optional<GetCustomerWSResponse> getCustomer(final String customerId) {
    LOGGER.debug("Calling webservice GET {}/{}", configuration.getUrl(), customerId);
    final URI query = fromHttpUrl(configuration.getUrl()).path("/").pathSegment(customerId)
        .build(true)
        .toUri();
    try {
      ResponseEntity<GetCustomerWSResponse> response = customerWebClient
          .getForEntity(query, GetCustomerWSResponse.class);
      LOGGER.debug("Web service GET {}/{} reply is : {}", configuration.getUrl(), customerId, response.getStatusCode());

      // on pourrait mettre un interceptor et ça serait mieux mais pour comparer c'est plus simple :P
      switch (response.getStatusCode()) {
        case OK:
          return Optional.ofNullable(response.getBody());
        case NOT_FOUND:
          return Optional.empty();
        default:
          Object body = Optional.ofNullable((Object) response.getBody())
              .orElse("");
          throw new WebServiceException("CUSTOMER_WS_GET_CUSTOMER_ERROR",
              "GET CustomerWS", response.getStatusCode(),
              "Unexpected response from the server while retrieving customer for customerId="
                  + customerId + ", response=" + body);
      }
    } catch (Exception err) {
      LOGGER.error("CustomerWS getCustomer failed", err);
      throw err;
    }
  }

}

File: customer-kotlin-coroutines-controllers/src/main/kotlin/com/prez/ws/CustomerWSClient.kt
package com.prez.ws

import com.prez.ws.model.GetCustomerWSResponse
import org.slf4j.LoggerFactory
import org.springframework.cloud.sleuth.annotation.NewSpan
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR
import org.springframework.http.HttpStatus.NOT_FOUND
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.WebClientResponseException
import org.springframework.web.reactive.function.client.awaitBody
import org.springframework.web.util.UriComponentsBuilder.fromHttpUrl

interface CustomerClient {
  suspend fun getCustomer(customerId: String): GetCustomerWSResponse?
}

open class CustomerWSClient(
  private val configuration: CustomerWSProperties,
  private val webClient: WebClient
) : CustomerClient {

  companion object {
    private val logger = LoggerFactory.getLogger(CustomerWSClient::class.java)
    private const val webServiceName = "CUSTOMER_WS"
  }

  @NewSpan("getCustomer")
  override suspend fun getCustomer(customerId: String): GetCustomerWSResponse? {
    logger.debug("Calling webservice GET ${configuration.url}/$customerId")
    val query = fromHttpUrl(configuration.url).path("/").pathSegment(customerId)
      .build(true)
      .toUri()

    try {
      val responseGet: GetCustomerWSResponse = webClient
        .get()
        .uri(query)
        .retrieve()
        .onStatus(HttpStatus::is3xxRedirection) { it.createException() } // Gestion du cas 3xx
        .awaitBody()
      logger.debug("GET CustomerWS retrieved customer : $responseGet")
      return responseGet
    } catch (e: Exception) {
      // on pourrait mettre une exchange filter function ici et ça serait mieux, mais c'est plus simple pour comparer :P
      logger.error("CustomerWS getCustomer failed", e)
      when { // and clearly you would love a guard condition :Pi
        e is WebClientResponseException && e.statusCode == NOT_FOUND -> return null
        e is WebClientResponseException -> throw WebServiceException(
          "CUSTOMER_WS_GET_CUSTOMER_ERROR", webServiceName, e.statusCode,
          "Unexpected response from the server while retrieving customer for customerId=$customerId, response=${e.responseBodyAsString}"
        )
        else -> throw WebServiceException(
          "CUSTOMER_WS_GET_CUSTOMER_ERROR",
          webServiceName,
          INTERNAL_SERVER_ERROR,
          "Unexpected error : ${e.message} for customerId=$customerId"
        )
      }
    }
  }
}